components/raftstore/src/store/fsm/store.rs

```diff
@@ -828,28 +800,72 @@ impl<EK: KvEngine, ER: RaftEngine, T: Transport> PollHandler<PeerFsm<EK, ER>, St
     }
 
     fn end(&mut self, peers: &mut [Box<PeerFsm<EK, ER>>]) {
-        self.flush_ticks();
+        for peer in peers {
+            peer.update_memory_trace(&mut self.trace_event);
+        }
+
+        let now = TiInstant::now();
+        if self.poll_ctx.trans.need_flush()
+            && now.saturating_duration_since(self.last_flush_msg_time)
+                >= Duration::from_micros(self.poll_ctx.cfg.raft_msg_flush_interval_us)
+        {
+            self.last_flush_msg_time = now;
+            self.poll_ctx.trans.flush();
+        }
+
+        let dur = now.saturating_duration_since(self.timer);
         if self.poll_ctx.has_ready {
-            self.handle_raft_ready(peers);
+            // Only enable the fail point when the store id is equal to 3, which is
+            // the id of slow store in tests.
+            fail_point!("on_raft_ready", self.poll_ctx.store_id() == 3, |_| {});
+
+            if !self.poll_ctx.store_stat.is_busy {
+                let election_timeout = Duration::from_millis(
+                    self.poll_ctx.cfg.raft_base_tick_interval.as_millis()
+                        * self.poll_ctx.cfg.raft_election_timeout_ticks as u64,
+                );
+                if dur >= election_timeout {
+                    self.poll_ctx.store_stat.is_busy = true;
+                }
+            }
+
+            slow_log!(
+                dur,
+                "{} handle {} pending peers include {} ready, {} entries, {} messages and {} \
+                 snapshots",
+                self.tag,
+                self.poll_ctx.pending_count,
+                self.poll_ctx.ready_count,
+                self.poll_ctx.raft_metrics.ready.append - self.previous_metrics.append,
+                self.poll_ctx.raft_metrics.ready.message - self.previous_metrics.message,
+                self.poll_ctx.raft_metrics.ready.snapshot - self.previous_metrics.snapshot
+            );
         }
+
         self.poll_ctx.current_time = None;
         self.poll_ctx
             .raft_metrics
             .process_ready
-            .observe(duration_to_sec(self.timer.saturating_elapsed()) as f64);
-        self.poll_ctx.raft_metrics.flush();
-        self.poll_ctx.store_stat.flush();
+            .observe(duration_to_sec(dur));
 
-        for peer in peers {
-            peer.update_memory_trace(&mut self.trace_event);
+        if now.saturating_duration_since(self.last_flush_time)
+            >= self.poll_ctx.cfg.store_events_flush_interval.0
```

Comment:
add a config check to make sure `store_events_flush_interval` is smaller than all ticks in `flush_ticks`?

Reply:
I remove this config and change it to 1ms.